Adding openfoam data for testing

mahmudulh-tamim opened the pull request:
I have put a template case in the `OF_cases/td_heat_conduction` directory. I also added the bash scripts for running the simulation several times based on the `template`.

#1

ragusa commented:
it is not reasonable to create a PR with over 3,000 files. github stalls when there are so many files. I was able to see fields that we do not care about, like gradTx, gradTy, gradTx, DT, and flux. These should be deleted, as we do not use them.

mahmudulh-tamim commented:
The bash scripts I used are case specific that is they are there to create new OF cases similar to the template and change the parameter DT in those cases,  and run simulation based on new DT values. Should I put them in the `examples/td_heat_conduction/bash` directory instead of in the `src/bash`? Moreover, should I create `examples/td_heat_conduction/python` directory and put the driver in that directory?
@ragusa

ragusa commented:
you can copy paste the bash script(s) here, in the dialog. you should also ping @rokkrpan for his input

ragusa commented:
i have looked at the 2 bash scripts you have placed in `src`. This is not correct.
1. first, `src/` is only for data assimilation, not managing the generation/cleaning of OF cases. If there are files that pertain to generating the snapshots, they must go in the `OF_cases` folder(s)
2. I disagree that you cannot make them general. This only thing that is hardcoded is the location of the cases. This must be an argument variable to the bash script, not hardcoded inside the script

mahmudulh-tamim commented:
let say I make them general. Then I should put them in `OF_cases/bash` folder?

ragusa commented:
yes, and update the readme over there to discuss what is found in that folder

mahmudulh-tamim commented:
I think that this PR solves the #1 for now. @ragusa @rokkrpan

rokkrpan commented:
In general, a typical OpenFoam user will firstly lookup for `Allrun` script.  

- Folders like `Data` could be omitted. Instead, prepare `DT_values` and 'file-names' in the `Allrun` script, located in the root folder. This also allows you to remove the `bash` folders. 
- Make scripts more general: replace DT with a generic variable, and specify the value, interval, how many steps,...
- Try to avoid jumping in and out of files, when running the code. When you run blockMesh run the solver as well. 
- Use native OpenFoam functions, like foamDictionary instead of sed. Pay attention to source the correct file in the header.

rokkrpan commented:
From what I saw yesterday, I am a bit afraid that things are getting messy. There are multiple scripts, each with a loop which does the same thing. There are driver_simulation, fix-space and runNtimes, which they use also a case specific files (`set-up-case`) - the same while loop is run three times, you add a few lines to set-up-case, and you do everything with a single file.

Next, there is already an "established" OpenFoam nomenclature, which is used in all repositories based on OpenFoam (OpenFoam.org, OpenFoam.com, GeN-Foam, DAFOAM,...). Such nomenclature is widely used  by the OpenFoam community, and it makes things easier with an experienced OpenFoam users. Namely, all the scripts which are used to run cases are named `Allrun`, and they are placed in the main folders. Having them in the `bash` folder named idontknowwhat along with 5 other scripts...

Here are a few examples. To run `td_heat_conduction/one-region-uniform-DT` you need (named `Allrun` placed in `td_heat_conduction/one-region-uniform-DT` ):
```
#!/bin/bash

#==============================================================================*

# Source tutorial run functions
. $WM_PROJECT_DIR/bin/tools/RunFunctions

#==============================================================================*

# OpenFoam template location
template_dir="./OpenFoamTemplate"

# List of all considered cases with variable values
file_cases="cases.txt"

# File where the variable is written
file_of_concern="constant/transportProperties"

#==============================================================================*

# Save current directory
current_dir=$(pwd)

# Generate mesh
echo "Generating mesh..."
blockMesh -case $template_dir > $template_dir/blockmesh.log

# Copy the template case into new case, change DT value and run the case.
while read -r caseName DT; do
    echo "Setting up" $caseName "..."
    mkdir -p $caseName
    cp -r $template_dir/* $caseName
    foamDictionary $caseName/${file_of_concern} -entry DT -set ${DT} > /dev/null 2>&1
    echo "Running..."
    cd $caseName
    $(getApplication) > run.log 2>&1
    cd $current_dir
done < <(tail -n +2 "$file_cases")

echo -e "Done! \n"
```

To clean `td_heat_conduction/one-region-uniform-DT` you need (named `Allclean`, `td_heat_conduction/one-region-uniform-DT` ):

```
#!/bin/bash

template_dir="./OpenFoamTemplate"
file_cases="cases.txt"

echo "Cleaning..."

rm -r $template_dir/constant/polyMesh > /dev/null 2>&1
rm -r $template_dir/*.log > /dev/null 2>&1

while read -r caseName DT; do
	rm -r $caseName > /dev/null 2>&1
done < <(tail -n +2 "$file_cases")

echo -e "Done! \n"
```

To run `td_heat_conduction/three-region-heteregenous-DT` you need (named `Allrun` placed in `td_heat_conduction/hree-region-heteregenous-DT` ):

```
#!/bin/bash

#==============================================================================*

# Source tutorial run functions
. $WM_PROJECT_DIR/bin/tools/RunFunctions

#==============================================================================*

# OpenFoam template location
template_dir="./OpenFoamTemplate"

# List of all considered cases with variable values
file_cases="cases.txt"

# File where the variable is written
file_of_concern="system/setFieldsDict"

#==============================================================================*

# Save current directory
current_dir=$(pwd)

# Generate mesh
echo "Generating mesh..."
blockMesh -case $template_dir > $template_dir/blockmesh.log

# Copy the template case into new case, change DT valu and run the case.
while read -r caseName DT; do
    echo "Setting up" $caseName "..."
    mkdir -p $caseName
    cp -r $template_dir/* $caseName
    sed -i "s/DT value_of_concern/DT ${DT}/g" $caseName/${file_of_concern}
    setFields -case $caseName > $caseName/setField.log 2>&1
    echo "Running..."
    cd $caseName
    $(getApplication) > run.log 2>&1
    cd $current_dir
done < <(tail -n +2 "$file_cases")

echo -e "Done! \n"
```

To clean it the same file as before (`Allclean` placed in `td_heat_conduction/hree-region-heteregenous-DT` ).

If you would like to run all the cases, then you can have `Allrun` in `OF_cases` (or whatever is the latest version of the folder with tests/examples):
```
# Use this script to run all cases in the subdirectories

find . -mindepth 2 -type f -name "Allrun" -execdir sh -c 'echo "Running: $(pwd)"; ./Allrun' sh {} \;
```

And `Allclean`
```
# Use this script to run all 'Allclean' in the subdirectories

find . -mindepth 2 -type f -name "Allclean" -execdir sh -c 'echo "Cleaning: $(pwd)"; ./Allclean' sh {} \; > /dev/null 2>$!
```
And, off coarse these scripts can be made more generic. 

You also need `cases.txt` file:

```
# Case name; DT value
A	2.848763e-04
B	3.158278e-04
C	1.432839e-04
D	1.398986e-04
E	2.198330e-04
F	2.318443e-04
G	2.753311e-04
H	3.210171e-04
I	1.176373e-04
J	3.363760e-04
```

Furthermore, naming consistency: `td_heat_conduction/one-region-uniform-DT`  - why are in one name use underscores and in the other one dashes?

mahmudulh-tamim commented:
I have combined all the bash scripts that were used to set case directory and run simulations into one `Allrun` script in each case-set's directory. Similary, I have added an `Allclean` script in the same directory. Furthermore, I have added in `examples/td_heat_conduction` another set of `Allrun` and `Allclean` to run all of  the `Allrun` and `Allclean` scripts inside the child directories, respectively. The child directories inside the `td_heat_conduction` are named with hyphens to make them distinguishable as a case-set directory in contrast to being a parent directory for a set of case-set directories.
Our next example could be `1D_MSR_pointKinetics`. Therefore, I have added the directory with just a readme.md.  
I hope that this pull request gets approved promptly, allowing me to begin working on issue #3  based on the updated main branch that results from merging this PR with the current main. However, if we need more minor changes, we can do that in next PRs.

ragusa commented on examples/td_heat_conduction/triZone-oneDynamicDT-Plate/Data/readme.md:
```diff
@@ -0,0 +1 @@
+This directory contains files and data related to the varying parameter.
```
since this reamd in inside a specific case, can the varying parameter be described with more details?

ragusa commented on examples/td_heat_conduction/triZone-plate/Allclean:
```diff
@@ -0,0 +1,22 @@
+# ensure that the script exits immediately if a command within exits with a non-zero status
+set -e
+
+#define the Data file
+#DataFile="./Data/data.csv"
+
```
delete commented out lines. why are you keeping that?

ragusa commented on examples/td_heat_conduction/triZone-plate/Allclean:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+#prepared by Mahmudul Hasan Tamim (mht@tamu.edu)
```
delete

ragusa commented on examples/td_heat_conduction/triZone-plate/Allclean:
```diff
@@ -0,0 +1,22 @@
+#DataFile="./Data/data.csv"
+
+# define the directory where different cases are kept
+case_space="./case_space"
```
why is the variable hardcoded?

mahmudulh-tamim replied:
same comment as above

ragusa commented on examples/td_heat_conduction/triZone-plate/Allclean:
```diff
@@ -0,0 +1,22 @@
+else
+   echo "${case_space} doesn't exist."
+fi
+
+
+
```
delete blank lines

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+#prepared by Mahmudul Hasan Tamim (mht@tamu.edu)
```
delete

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+}
+
+#define the Data file
+DataFile="./Data/data.csv"
```
why hardcoded variable?

mahmudulh-tamim replied:
same comment as above

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+case_space="./case_space"
+
+# define the template directory
+
```
delete blank line

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+fi
+
+# checks if the template directory exists
+
```
delete blank line

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+# checks if the template directory exists
+
+if [[ ! -d ${directory_template} ]]; then
+
```
delete blank line

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+fi
+
+if [[ ! -d ${case_space} ]]; then
+
```
delete blank line

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+if [[ ! -d ${case_space} ]]; then
+
+  mkdir ${case_space}
+ 
```
delete blank line

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+fi
+
+#read the header line
+
```
delete commented out lines and excess of blank lines

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+
+
+# read data row by row
+
```
delete blank line

ragusa commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+	${applicationName} > case-run.log 2>&1
+    	echo "simulation done in ${new_case}"
+    )
+    
+    
```
delete blank lines

ragusa commented on examples/td_heat_conduction/triZone-plate/Data/readme.md:
```diff
@@ -0,0 +1 @@
+This directory contains files and data related to the varying parameter.
```
can you be more explicit about the parameter

ragusa commented on examples/td_heat_conduction/uniformDT-plate/Allrun:
same comments as before

ragusa commented on examples/td_heat_conduction/uniformDT-plate/Allclean:
same comments as before

mahmudulh-tamim commented on examples/td_heat_conduction/triZone-oneDynamicDT-Plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+}
+
+#define the Data file
+DataFile="./Data/data.csv"
```
I hardcoded it because the `Allrun` script assumes a certain directory structure, and, moreover, `Allrun` scripts usually don't take any argument.

rokkrpan replied:
`Allrun` are usually made for specific case, so no additional arguments are needed to run them.

Regarding this particular `Allrun`...well, the checks are not needed (mostly because bash already reveals the reason for the error), the definition of 'set-up-case' function is again case specific (plus a check) and could be replaced with two lines in the loop below:
```
# Set thermal diffusivity
sed -i "s/volScalarFieldValue DT 0.04/volScalarFieldValue DT ${DT}/g" system/setFieldsDict
```

mahmudulh-tamim replied:
I disagree that the scripts should not perform defensive checks. I would move for more checks and more thorough ones. However, given that it makes progress slow, we can put some precursors to those more rigorous checks. And these checks were part of this plan. 
For example, the data.csv file should not only just be checked for existence, but also should be checked for if the data is in correct format, etc.

mahmudulh-tamim replied:
The `set-up-case` function is compilation of the case-set specific commands, that is why I kept it separate for each case-set's `Allrun` script for more modularity.

rokkrpan replied:
@mahmudulh-tamim,  `cp`, `rm` and `mkdir` bash functions will already give you the 'No such file or directory' output in the case of non existent files. And this is why I said in one of my previous comments 'obvious ChatGPT', which always enforces some not needed checks with the same output.  When using `set -e` option (as you already did), the first encountered error of any of these function, will stop the further execution. These type of checks are indeed needed if you do not want to rewrite the files, but the intention of this scripts is to generate the snapshots on a empty folder.

The story on the `data.csv` file is a bit different, since CSV stands for comma separated values.

mahmudulh-tamim replied:
I understand,but I am going to keep those checks for now.

mahmudulh-tamim commented on examples/td_heat_conduction/triZone-oneDynamicDT-Plate/Allclean:
```diff
@@ -0,0 +1,21 @@
+#DataFile="./Data/data.csv"
+
+# define the directory where different cases are kept
+case_space="./case_space"
```
same comment as above

mahmudulh-tamim commented on OF_cases/td_heat_conduction/three-region-heterogenous-DT/bash/set-up-case:
I think you are referring to a previous commit @rokkrpan .

mahmudulh-tamim commented on OF_cases/td_heat_conduction/three-region-heterogenous-DT/bash/set-up-case:
```diff
@@ -0,0 +1,32 @@
+
+	cd "${nom}"
+	sed -i "s/DT 0.0004/DT ${DT}/g" ${file_of_concern}
+	blockMesh > case-blockmesh.log 2>&1
```
It's from a previous commit.

mahmudulh-tamim commented on OF_cases/td_heat_conduction/three-region-heterogenous-DT/names-cases.txt:
You are referring to a previous commit. In the latest commit, for which I have asked for your review, implemented most of your recommendations.

mahmudulh-tamim commented on examples/td_heat_conduction/triZone-plate/Allrun:
```diff
@@ -0,0 +1,82 @@
+        echo "Mesh has been set up."
+        setFields > case-setFields.log 2>&1
+        applicationName=$(foamDictionary -entry application -value system/controlDict)
+	${applicationName} > case-run.log 2>&1
```
it's the same, right? Because the $(getApplication) runs the same command and I learned it from you.

mahmudulh-tamim replied:
Done

mahmudulh-tamim commented on examples/readme.md:
```diff
@@ -1,4 +1,8 @@
+**[td_heat_conduction](./td_heat_conduction)**  contains the case sets for the different variations of the time dependent heat conduction. 
+
+**[1D_MSR_pointKinetics](./1D_MSR_pointKinetics)** contains the case sets for the different variations of the 1D MSR. 
+
```
Done.

mahmudulh-tamim commented:
I have done most of the requested changes and resolved associated conversations. The conversations I haven't resolved are because I am waiting for your further comments.

ragusa commented:
i can approve when the unresolved discussion are satisfactorily resolved. I think my point is that I do not like to have files with identical filenames but that are doing different things. It looks like @rokkrpan posted a script. I think this is the last thing on my side. You state that you are waiting for further comments, but Rok did leave comments. I do not follow that last bit of needing further comments

mahmudulh-tamim commented on examples/td_heat_conduction/Allclean:
I have deleted the `Allrun` (`driverAllrun`) in the directory above and changed the `Allclean` (`driverAllclean`) to your recommended code.
I am going to resolve the conversations soon.

rokkrpan commented:
@mahmudulh-tamim , I know that I already approved this, (and since there are only folders in there) you should add back the 1D MSR case in order to compare the results with the ones from Francesco (and the sensor placement can be also easily checked with the intuition).

mahmudulh-tamim commented:
1D MSR could be added; therefore, I added it now. The previous one was 1D MSR Point Kinetics and I thought that example was not an optimal choice for a demonstration of application of GEIM; therefore, I deleted it. However, currently, I am working on 2D MSFR.
